# Week 2 - Linear Regression 2

In [1]:
#pip install kaggle

In [2]:
import os
import json

# Replace this with the content of your kaggle.json
kaggle_token = {
    "username": "your_kaggle_username",
    "key": "your_kaggle_api_key"
}

os.makedirs(os.path.expanduser("~/.kaggle"), exist_ok=True)
with open(os.path.expanduser("~/.kaggle/kaggle.json"), "w") as f:
    json.dump(kaggle_token, f)

os.chmod(os.path.expanduser("~/.kaggle/kaggle.json"), 0o600)
import os

# Make sure your kaggle.json is correctly set up
os.environ['KAGGLE_CONFIG_DIR'] = os.path.expanduser("~/.kaggle")

# Download the dataset
!kaggle datasets download -d ahmedshahriarsakib/usa-real-estate-dataset

# Unzip the dataset
!unzip -o usa-real-estate-dataset.zip -d usa_real_estate



Dataset URL: https://www.kaggle.com/datasets/ahmedshahriarsakib/usa-real-estate-dataset
License(s): other
  0%|                                               | 0.00/38.2M [00:00<?, ?B/s]
100%|██████████████████████████████████████| 38.2M/38.2M [00:00<00:00, 1.14GB/s]
Archive:  usa-real-estate-dataset.zip
  inflating: usa_real_estate/realtor-data.zip.csv  


In [3]:
import pandas as pd

df = pd.read_csv('usa_real_estate/realtor-data.zip.csv')  
print(df.head())

   brokered_by    status     price  bed  bath  acre_lot     street  \
0     103378.0  for_sale  105000.0  3.0   2.0      0.12  1962661.0   
1      52707.0  for_sale   80000.0  4.0   2.0      0.08  1902874.0   
2     103379.0  for_sale   67000.0  2.0   1.0      0.15  1404990.0   
3      31239.0  for_sale  145000.0  4.0   2.0      0.10  1947675.0   
4      34632.0  for_sale   65000.0  6.0   2.0      0.05   331151.0   

         city        state  zip_code  house_size prev_sold_date  
0    Adjuntas  Puerto Rico     601.0       920.0            NaN  
1    Adjuntas  Puerto Rico     601.0      1527.0            NaN  
2  Juana Diaz  Puerto Rico     795.0       748.0            NaN  
3       Ponce  Puerto Rico     731.0      1800.0            NaN  
4    Mayaguez  Puerto Rico     680.0         NaN            NaN  


In [4]:
austin_df = df[(df['city'] == 'Austin') & (df['state'] == 'Texas')]
austin_df = austin_df[['price', 'bed', 'bath', 'acre_lot', 'house_size']].dropna()
austin_df

price  bed  bath  acre_lot  house_size
1119516   500000.0  3.0   2.0      0.16      1320.0
1119518   650000.0  2.0   2.0      0.09       870.0
1119522   350000.0  2.0   1.0      0.03       873.0
1119523  1345000.0  4.0   4.0      0.36      3496.0
1119524   599000.0  3.0   2.0      0.16      1484.0
...            ...  ...   ...       ...         ...
1991922   559000.0  6.0   4.0      0.36      2927.0
1991923   579000.0  3.0   2.0      0.22      1465.0
1991924   830000.0  4.0   3.0      0.21      3220.0
1991925   399900.0  3.0   2.0      0.18      1505.0
1991926  2199000.0  4.0   5.0      1.00      4731.0

[1512 rows x 5 columns]

In [5]:
from sklearn.preprocessing import StandardScaler
X = austin_df[['bed', 'bath', 'acre_lot', 'house_size']]
y = austin_df['price']

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

Try Ridge Model

In [6]:
from sklearn.linear_model import Ridge, Lasso, ElasticNet
from sklearn.model_selection import cross_val_score
import numpy as np

# Ridge Regression
ridge = Ridge(alpha=1.0)
ridge_score = np.mean(cross_val_score(ridge, X_scaled, y, scoring='r2', cv=5))
ridge_score

np.float64(0.529034417204002)

In this analysis, I implemented Ridge Regression, a type of linear regression that includes L2 regularization to prevent overfitting. The model was trained to predict housing prices in Austin, Texas. The R² score of 0.5290 indicates that approximately 52.9% of the variance in housing prices is explained by the model.This suggests a moderate level of predictive power—the model captures general pricing trends but may miss other complex factors

Try Lasso

In [7]:
# Lasso Regression
lasso = Lasso(alpha=0.1)
lasso_score = np.mean(cross_val_score(lasso, X_scaled, y, scoring='r2', cv=5))
lasso_score

np.float64(0.5285799559925718)

In this analysis, I applied Lasso Regression, which extends linear regression with L1 regularization.The model achieved an R² score of 0.52857. This means that the model explains approximately 52.86% of the variability in housing prices. 

Try Elastic Net regression

In [8]:
# Elastic Net Regression
elastic = ElasticNet(alpha=0.1, l1_ratio=0.5)
elastic_score = np.mean(cross_val_score(elastic, X_scaled, y, scoring='r2', cv=5))
elastic_score

np.float64(0.5484837985529772)

The Elastic Net regression model combines the strengths of Ridge (L2) and Lasso (L1) regularization by introducing a penalty that is a mix of both. The resulting R² score was 0.5484, indicating that approximately 54.84% of the variance in housing prices was explained by the model. This represents the highest predictive performance among the three regularized models tested (Ridge = 0.5290, Lasso = 0.5286).

The results suggest that Elastic Net regression is the most suitable for this housing dataset, as it handles correlated predictors while offering improved prediction accuracy. 